

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Preprocessing iCLIP/eCLIP-seq reads &mdash; PureCLIP v1.0.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="PureCLIP v1.0.0 documentation" href="../index.html"/>
        <link rel="up" title="Getting Started" href="index.html"/>
        <link rel="next" title="PureCLIP Tutorial" href="../PureCLIPTutorial/index.html"/>
        <link rel="prev" title="Requirements" href="requirements.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> PureCLIP
          

          
          </a>

          
            
            
              <div class="version">
                v1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Getting Started</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="requirements.html">Requirements</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Preprocessing iCLIP/eCLIP-seq reads</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../PureCLIPTutorial/index.html">PureCLIP Tutorial</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">PureCLIP</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Getting Started</a> &raquo;</li>
        
      <li>Preprocessing iCLIP/eCLIP-seq reads</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/GettingStarted/preprocessing.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="preprocessing-iclip-eclip-seq-reads">
<h1>Preprocessing iCLIP/eCLIP-seq reads<a class="headerlink" href="#preprocessing-iclip-eclip-seq-reads" title="Permalink to this headline">¶</a></h1>
<p>This tutorial will walk you through an example how to preprocess your iCLIP/eCLIP data before analysing it.
This only serves as a guide and you can of course as well use your own CLIP-seq preprocessing workflow.</p>
<div class="section" id="requirements-for-this-tutorial">
<h2>Requirements for this tutorial<a class="headerlink" href="#requirements-for-this-tutorial" title="Permalink to this headline">¶</a></h2>
<p>The tutorial requires the following tools (and was tested for the specified versions):</p>
<ul class="simple">
<li><a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/installation.html">cutadapt</a> (v1.12)</li>
<li><a class="reference external" href="https://github.com/samtools/samtools">samtools</a> (v0.1.19-44428cd)</li>
<li><a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> (v2.5.1b)</li>
<li><a class="reference external" href="https://github.com/CGATOxford/UMI-tools">UMI</a> (v0.4.3)</li>
<li><a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastqc</a> (v0.11.5)</li>
</ul>
</div>
<div class="section" id="data-retrieval">
<h2>Data retrieval<a class="headerlink" href="#data-retrieval" title="Permalink to this headline">¶</a></h2>
<p>Let’s start with downloading the data.
We use the raw <a class="reference external" href="https://www.encodeproject.org/experiments/ENCSR661ICQ/">PUM2 eCLIP sequencing data</a> from ENCODE (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/27018577">Van Nostrand et. al, 2016</a>):</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">rep1</span> <span class="n">rep2</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">www</span><span class="o">.</span><span class="n">encodeproject</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">ENCFF956TOZ</span><span class="o">/@</span><span class="nd">@download</span><span class="o">/</span><span class="n">ENCFF956TOZ</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">www</span><span class="o">.</span><span class="n">encodeproject</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">ENCFF133DNM</span><span class="o">/@</span><span class="nd">@download</span><span class="o">/</span><span class="n">ENCFF133DNM</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">rep2</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">www</span><span class="o">.</span><span class="n">encodeproject</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">ENCFF041KJT</span><span class="o">/@</span><span class="nd">@download</span><span class="o">/</span><span class="n">ENCFF041KJT</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">rep2</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">www</span><span class="o">.</span><span class="n">encodeproject</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">ENCFF462SCV</span><span class="o">/@</span><span class="nd">@download</span><span class="o">/</span><span class="n">ENCFF462SCV</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>We download the human reference genome (primary assembly containing chromosomes and scaffolds) and the genome annotation (comprehensive set) from <a class="reference external" href="http://www.gencodegenes.org/">GENCODE</a>:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">ref</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sanger</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">pub</span><span class="o">/</span><span class="n">gencode</span><span class="o">/</span><span class="n">Gencode_human</span><span class="o">/</span><span class="n">release_26</span><span class="o">/</span><span class="n">GRCh38</span><span class="o">.</span><span class="n">primary_assembly</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>
<span class="n">gunzip</span> <span class="n">ref</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>

<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">annotation</span><span class="o">.</span><span class="n">v26</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sanger</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">pub</span><span class="o">/</span><span class="n">gencode</span><span class="o">/</span><span class="n">Gencode_human</span><span class="o">/</span><span class="n">release_26</span><span class="o">/</span><span class="n">gencode</span><span class="o">.</span><span class="n">v26</span><span class="o">.</span><span class="n">primary_assembly</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span>
<span class="n">gunzip</span> <span class="n">annotation</span><span class="o">.</span><span class="n">v26</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Please note, that the following steps need to be done for both replicates, while only being described for rep1.</p>
</div>
<div class="section" id="adaptor-trimming">
<h2>Adaptor trimming<a class="headerlink" href="#adaptor-trimming" title="Permalink to this headline">¶</a></h2>
<p>Possible adapter contaminations at 3’ ends can be removed using the tool <a class="reference external" href="https://github.com/marcelm/cutadapt">cutadapt</a> <a class="reference external" href="http://journal.embnet.org/index.php/embnetjournal/article/view/200">(Martin, 2011)</a> (as done by the YEO lab in the ENCODE <a class="reference external" href="https://www.encodeproject.org/documents/dde0b669-0909-4f8b-946d-3cb9f35a6c52/&#64;&#64;download/attachment/eCLIP_analysisSOP_v1.P.pdf">processing pipeline</a>), while discarding reads shorter than 18bp:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">cutadapt</span> <span class="o">-</span><span class="n">f</span> <span class="n">fastq</span> <span class="o">--</span><span class="n">match</span><span class="o">-</span><span class="n">read</span><span class="o">-</span><span class="n">wildcards</span> <span class="o">--</span><span class="n">times</span> <span class="mi">1</span> <span class="o">-</span><span class="n">e</span> <span class="mf">0.1</span> <span class="o">-</span><span class="n">O</span> <span class="mi">1</span> <span class="o">--</span><span class="n">quality</span><span class="o">-</span><span class="n">cutoff</span> <span class="mi">6</span> <span class="o">-</span><span class="n">m</span> <span class="mi">18</span> <span class="o">-</span><span class="n">a</span> <span class="n">NNNNNAGATCGGAAGAGCACACGTCTGAACTCCAGTCAC</span> <span class="o">-</span><span class="n">g</span> <span class="n">CTTCCGATCTACAAGTT</span>  <span class="o">-</span><span class="n">g</span> <span class="n">CTTCCGATCTTGGTCCT</span> <span class="o">-</span><span class="n">A</span> <span class="n">AACTTGTAGATCGGA</span> <span class="o">-</span><span class="n">A</span> <span class="n">AGGACCAAGATCGGA</span> <span class="o">-</span><span class="n">A</span> <span class="n">ACTTGTAGATCGGAA</span> <span class="o">-</span><span class="n">A</span> <span class="n">GGACCAAGATCGGAA</span>  <span class="o">-</span><span class="n">A</span> <span class="n">CTTGTAGATCGGAAG</span>  <span class="o">-</span><span class="n">A</span> <span class="n">GACCAAGATCGGAAG</span> <span class="o">-</span><span class="n">A</span> <span class="n">TTGTAGATCGGAAGA</span> <span class="o">-</span><span class="n">A</span> <span class="n">ACCAAGATCGGAAGA</span> <span class="o">-</span><span class="n">A</span> <span class="n">TGTAGATCGGAAGAG</span> <span class="o">-</span><span class="n">A</span> <span class="n">CCAAGATCGGAAGAG</span> <span class="o">-</span><span class="n">A</span> <span class="n">GTAGATCGGAAGAGC</span> <span class="o">-</span><span class="n">A</span> <span class="n">CAAGATCGGAAGAGC</span> <span class="o">-</span><span class="n">A</span> <span class="n">TAGATCGGAAGAGCG</span>  <span class="o">-</span><span class="n">A</span> <span class="n">AAGATCGGAAGAGCG</span> <span class="o">-</span><span class="n">A</span> <span class="n">AGATCGGAAGAGCGT</span>  <span class="o">-</span><span class="n">A</span> <span class="n">GATCGGAAGAGCGTC</span> <span class="o">-</span><span class="n">A</span> <span class="n">ATCGGAAGAGCGTCG</span> <span class="o">-</span><span class="n">A</span> <span class="n">TCGGAAGAGCGTCGT</span> <span class="o">-</span><span class="n">A</span> <span class="n">CGGAAGAGCGTCGTG</span> <span class="o">-</span><span class="n">A</span> <span class="n">GGAAGAGCGTCGTGT</span> <span class="o">-</span><span class="n">o</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">trimmed</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">p</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">trimmed</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>For eCLIP data, it was suggested (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/27018577">Van Nostrand et. al, 2016</a>) to apply two rounds of adapter trimming to correct for possible double ligations events.</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">cutadapt</span> <span class="o">-</span><span class="n">f</span> <span class="n">fastq</span> <span class="o">--</span><span class="n">match</span><span class="o">-</span><span class="n">read</span><span class="o">-</span><span class="n">wildcards</span> <span class="o">--</span><span class="n">times</span> <span class="mi">1</span> <span class="o">-</span><span class="n">e</span> <span class="mf">0.1</span> <span class="o">-</span><span class="n">O</span> <span class="mi">5</span> <span class="o">--</span><span class="n">quality</span><span class="o">-</span><span class="n">cutoff</span> <span class="mi">6</span> <span class="o">-</span><span class="n">m</span> <span class="mi">18</span> <span class="o">-</span><span class="n">A</span> <span class="n">AACTTGTAGATCGGA</span> <span class="o">-</span><span class="n">A</span> <span class="n">AGGACCAAGATCGGA</span> <span class="o">-</span><span class="n">A</span> <span class="n">ACTTGTAGATCGGAA</span> <span class="o">-</span><span class="n">A</span> <span class="n">GGACCAAGATCGGAA</span> <span class="o">-</span><span class="n">A</span> <span class="n">CTTGTAGATCGGAAG</span> <span class="o">-</span><span class="n">A</span> <span class="n">GACCAAGATCGGAAG</span> <span class="o">-</span><span class="n">A</span> <span class="n">TTGTAGATCGGAAGA</span> <span class="o">-</span><span class="n">A</span> <span class="n">ACCAAGATCGGAAGA</span> <span class="o">-</span><span class="n">A</span> <span class="n">TGTAGATCGGAAGAG</span> <span class="o">-</span><span class="n">A</span> <span class="n">CCAAGATCGGAAGAG</span> <span class="o">-</span><span class="n">A</span> <span class="n">GTAGATCGGAAGAGC</span> <span class="o">-</span><span class="n">A</span> <span class="n">CAAGATCGGAAGAGC</span> <span class="o">-</span><span class="n">A</span> <span class="n">TAGATCGGAAGAGCG</span> <span class="o">-</span><span class="n">A</span> <span class="n">AAGATCGGAAGAGCG</span> <span class="o">-</span><span class="n">A</span> <span class="n">AGATCGGAAGAGCGT</span> <span class="o">-</span><span class="n">A</span> <span class="n">GATCGGAAGAGCGTC</span> <span class="o">-</span><span class="n">A</span> <span class="n">ATCGGAAGAGCGTCG</span> <span class="o">-</span><span class="n">A</span> <span class="n">TCGGAAGAGCGTCGT</span> <span class="o">-</span><span class="n">A</span> <span class="n">CGGAAGAGCGTCGTG</span> <span class="o">-</span><span class="n">A</span> <span class="n">GGAAGAGCGTCGTGT</span> <span class="o">-</span><span class="n">o</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">p</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">trimmed</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">trimmed</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
<div class="section" id="preparing-read-ids-for-umi">
<h2>Preparing read IDs for UMI<a class="headerlink" href="#preparing-read-ids-for-umi" title="Permalink to this headline">¶</a></h2>
<p>We remove PCR duplicates based on the mapping position and random barcodes using UMI, which requires the read ID in the format <code class="docutils literal"><span class="pre">_&#64;HISEQ:87:00000000_BARCODE</span> <span class="pre">read1_</span></code>.
Therefore we append the barcode to the read ID prior the mapping, using the following command:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">zcat</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">&gt;</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">zcat</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">&gt;</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">awk</span> <span class="o">-</span><span class="n">v</span> <span class="n">l</span><span class="o">=</span><span class="mi">10</span> <span class="s1">&#39;BEGIN{OFS=FS=&quot; &quot;} substr($1, 1, 1) == &quot;@&quot; {print &quot;@&quot; substr($1, (l+3), 500) &quot;_&quot; substr($1, 2, l) &quot; &quot; $2 }; substr($1, 1, 1) != &quot;@&quot; </span><span class="si">{print}</span><span class="s1">; &#39;</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">fastq</span>  <span class="o">|</span> <span class="n">gzip</span> <span class="o">&gt;</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">bc</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">awk</span> <span class="o">-</span><span class="n">v</span> <span class="n">l</span><span class="o">=</span><span class="mi">10</span> <span class="s1">&#39;BEGIN{OFS=FS=&quot; &quot;} substr($1, 1, 1) == &quot;@&quot; {print &quot;@&quot; substr($1, (l+3), 500) &quot;_&quot; substr($1, 2, l) &quot; &quot; $2 }; substr($1, 1, 1) != &quot;@&quot; </span><span class="si">{print}</span><span class="s1">; &#39;</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">fastq</span>  <span class="o">|</span> <span class="n">gzip</span> <span class="o">&gt;</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">bc</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>where <code class="docutils literal"><span class="pre">l=10</span></code> denotes the used barcode length.</p>
</div>
<div class="section" id="read-mapping-with-star">
<h2>Read mapping with STAR<a class="headerlink" href="#read-mapping-with-star" title="Permalink to this headline">¶</a></h2>
<p>CLIP-seq reads can be mapped with the RNA-seq read aligner <a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/23104886">Dobin et. al, 2013</a>).
It allows to include genome annotations in order to enable the alignment against spliced transcripts.
First, we need to prepare the genome index, using the annotation file:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">genome_index</span>
<span class="n">STAR</span> <span class="o">--</span><span class="n">runThreadN</span> <span class="mi">10</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">genome_index</span><span class="o">/</span> <span class="o">--</span><span class="n">genomeFastaFiles</span> <span class="n">ref</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="n">fa</span> <span class="o">--</span><span class="n">sjdbGTFfile</span> <span class="n">annotation</span><span class="o">.</span><span class="n">v26</span><span class="o">.</span><span class="n">gtf</span> <span class="o">--</span><span class="n">sjdbOverhang</span> <span class="mi">49</span>
</pre></div>
</div>
<p>Next, we map the reads (R1 and R2) against the indexed genome:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="n">rep1</span><span class="o">/</span><span class="n">STAR</span>
<span class="n">STAR</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">runThreadN</span> <span class="mi">10</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">genome_index</span><span class="o">/</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">bc</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rep1</span><span class="o">/</span><span class="n">reads</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">trimmed2</span><span class="o">.</span><span class="n">bc</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">--</span><span class="n">readFilesCommand</span>  <span class="n">zcat</span> <span class="o">--</span><span class="n">outFilterType</span> <span class="n">BySJout</span> <span class="o">--</span><span class="n">outFilterMultimapNmax</span> <span class="mi">1</span> <span class="o">--</span><span class="n">alignSJoverhangMin</span> <span class="mi">8</span> <span class="o">--</span><span class="n">alignSJDBoverhangMin</span> <span class="mi">1</span> <span class="o">--</span><span class="n">outFilterMismatchNmax</span> <span class="mi">999</span> <span class="o">--</span><span class="n">outFilterMismatchNoverLmax</span> <span class="mf">0.04</span> <span class="o">--</span><span class="n">scoreDelOpen</span> <span class="o">-</span><span class="mi">1</span> <span class="o">--</span><span class="n">alignIntronMin</span> <span class="mi">20</span> <span class="o">--</span><span class="n">alignIntronMax</span> <span class="mi">1000000</span> <span class="o">--</span><span class="n">alignMatesGapMax</span> <span class="mi">1000000</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">rep1</span><span class="o">/</span><span class="n">STAR</span><span class="o">/</span> <span class="o">--</span><span class="n">alignEndsType</span> <span class="n">EndToEnd</span>
</pre></div>
</div>
<p>The parameter <code class="docutils literal"><span class="pre">--outFilterMultimapNmax</span> <span class="pre">1</span></code> ensures only uniquely mapping reads will be reported.
Since we used the primary assembly containing scaffolds as reference, this enables us to filter out reads that map both against a main chromosome and against a scaffold (e.g. ribosomal RNA).
Furthermore, it is important to use the <code class="docutils literal"><span class="pre">--alignEndsType</span> <span class="pre">EndToEnd</span></code> setting, to ensure the mapping of the whole read.
The aligned reads will be written then to STAR/Aligned.sortedByCoord.out.bam .</p>
</div>
<div class="section" id="filtering">
<h2>Filtering<a class="headerlink" href="#filtering" title="Permalink to this headline">¶</a></h2>
<p>We filter the aligned reads to obtain only reads mapping against the main chromosomes:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">index</span> <span class="n">rep1</span><span class="o">/</span><span class="n">STAR</span><span class="o">/</span><span class="n">Aligned</span><span class="o">.</span><span class="n">sortedByCoord</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">bam</span>
<span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">hb</span> <span class="o">-</span><span class="n">f</span> <span class="mi">2</span> <span class="n">rep1</span><span class="o">/</span><span class="n">STAR</span><span class="o">/</span><span class="n">Aligned</span><span class="o">.</span><span class="n">sortedByCoord</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">bam</span> <span class="o">-</span><span class="n">o</span> <span class="n">rep1</span><span class="o">/</span><span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">bam</span> <span class="n">chr1</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr2</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr3</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr4</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr5</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr6</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr7</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr8</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr9</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr10</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr11</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr12</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr13</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr14</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr15</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr16</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr17</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr18</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr19</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr20</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr21</span><span class="p">:</span><span class="mi">1</span> <span class="n">chr22</span><span class="p">:</span><span class="mi">1</span> <span class="n">chrX</span><span class="p">:</span><span class="mi">1</span> <span class="n">chrY</span><span class="p">:</span><span class="mi">1</span>
<span class="n">samtools</span> <span class="n">index</span> <span class="n">rep1</span><span class="o">/</span><span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</div>
<div class="section" id="pcr-duplicate-removal-using-umi">
<h2>PCR duplicate removal using UMI<a class="headerlink" href="#pcr-duplicate-removal-using-umi" title="Permalink to this headline">¶</a></h2>
<p>For truncation based CLIP-seq data it is crucial to remove PCR duplicates to allow for an accurate crosslink site detection.
We use the tool <a class="reference external" href="https://github.com/CGATOxford/UMI-tools">UMI</a> (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5340976/">Smith et. al, 2017</a>), which is able to handle errors within barcode sequences.</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">umi_tools</span> <span class="n">dedup</span> <span class="o">-</span><span class="n">I</span> <span class="n">rep1</span><span class="o">/</span><span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">paired</span> <span class="o">-</span><span class="n">S</span> <span class="n">rep1</span><span class="o">/</span><span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">duplRm</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</div>
<div class="section" id="pooling-and-r2-retrieval">
<h2>Pooling and R2 retrieval<a class="headerlink" href="#pooling-and-r2-retrieval" title="Permalink to this headline">¶</a></h2>
<p>Finally, we merge the preprocessed alignments of the individual replicates:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">merge</span> <span class="o">-</span><span class="n">f</span> <span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">duplRm</span><span class="o">.</span><span class="n">pooled</span><span class="o">.</span><span class="n">bam</span> <span class="n">rep1</span><span class="o">/</span><span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">duplRm</span><span class="o">.</span><span class="n">bam</span> <span class="n">rep2</span><span class="o">/</span><span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">duplRm</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>and filter for R2, to keep only reads containing information about potential truncation events (for iCLIP data this would be R1):</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">hb</span> <span class="o">-</span><span class="n">f</span> <span class="mi">130</span> <span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">duplRm</span><span class="o">.</span><span class="n">pooled</span><span class="o">.</span><span class="n">bam</span> <span class="o">-</span><span class="n">o</span> <span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">duplRm</span><span class="o">.</span><span class="n">pooled</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">bam</span>
<span class="n">samtools</span> <span class="n">index</span> <span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">duplRm</span><span class="o">.</span><span class="n">pooled</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</div>
<div class="section" id="quality-control">
<h2>Quality control<a class="headerlink" href="#quality-control" title="Permalink to this headline">¶</a></h2>
<p>It’s always a good idea to assess the quality of the data prior to the actual analysis.
For this we use <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastqc</a>:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">fastqc</span>
<span class="n">fastqc</span> <span class="o">-</span><span class="n">o</span> <span class="n">fastqc</span><span class="o">/</span> <span class="n">aligned</span><span class="o">.</span><span class="n">f</span><span class="o">.</span><span class="n">duplRm</span><span class="o">.</span><span class="n">pooled</span><span class="o">.</span><span class="n">R2</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../PureCLIPTutorial/index.html" class="btn btn-neutral float-right" title="PureCLIP Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="requirements.html" class="btn btn-neutral" title="Requirements" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Sabrina Krakau.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'v1.0.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
  
 <script type="text/javascript">
    $(document).ready(function() {
        $(".toggle > *").hide();
        $(".toggle .header").show();
        $(".toggle .header").click(function() {
            $(this).parent().children().not(".header").toggle(400);
            $(this).parent().children(".header").toggleClass("open");
        })
    });
</script>


</body>
</html>